### PR TITLE
Add some useful additional printer columns for TidbMonitor

### LINKED ISSUE #2169: Add some useful additional printer columns for TidbMonitor
## Feature Request

**Is your feature request related to a problem? Please describe:**

Now it uses the default columns.

```
$ kubectl get tidbmonitor
NAME    AGE
basic   23s
```

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve?
#3292
Fix #2169

### What is changed and how does it work?
Add some useful additional printer columns for TidbMonitor

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects
None.

### Release Notes

```release-note
Add some useful additional printer columns for TidbMonitor.
```

### CODE REVIEW

File: pkg/util/crdutil.go [hunk 1]
```diff
@@ -296,6 +296,24 @@ var (
...
+	readyTiDBMonitorAdditionalPrinterColumns = extensionsobj.CustomResourceColumnDefinition{
+		Name:     "READY",
+		Type:     "string",
+		JSONPath: `.status.statefulSet.conditions[?(@.type=="Ready")].status`,
+	}
+
+	tidbMonitorDesireColumn = extensionsobj.CustomResourceColumnDefinition{
+		Name:        "Desire",
+		Type:        "integer",
+		Description: "The desired replicas number of TiKV cluster",
```
Comment (csuzhangxc) [inline pkg/util/crdutil.go:309]: not `TiKV cluster` here.

File: pkg/util/crdutil.go [hunk 2]
```diff
@@ -296,6 +296,24 @@ var (
...
+		JSONPath:    ".spec.replicas",
+	}
+	statusTiDBMonitorAdditionalPrinterColumns = extensionsobj.CustomResourceColumnDefinition{
+		Name:     "UP-TO-DATE",
+		Type:     "string",
```
Comment (csuzhangxc) [inline pkg/util/crdutil.go:314]: Can we use `integer` here?

File: pkg/monitor/monitor/monitor_manager.go [hunk 3]
```diff
@@ -169,11 +169,31 @@ func (m *MonitorManager) SyncMonitor(monitor *v1alpha1.TidbMonitor) error {
...
 
 	return nil
 }
 
+func (m *MonitorManager) syncTidbMonitorStatus(monitor *v1alpha1.TidbMonitor) error {
+	sts, err := m.deps.StatefulSetLister.StatefulSets(monitor.Namespace).Get(GetMonitorObjectName(monitor))
+	if err != nil {
+		return err
+	}
+	if sts == nil {
```
Comment (july2993) [inline pkg/monitor/monitor/monitor_manager.go:189]: this is an unreachable branch

**Review by july2993 — COMMENTED** [review]
rest lgtm

could you show an example output?

File: manifests/crd.yaml [hunk 4]
```diff
@@ -19214,12 +19214,26 @@ metadata:
   creationTimestamp: null
   name: tidbmonitors.pingcap.com
 spec:
+  additionalPrinterColumns:
+    - JSONPath: .status.statefulSet.conditions[?(@.type=="Ready")].status
```
[the remark this answers is not in the document]
Comment (mikechengwei) [reply manifests/crd.yaml]: None , i’m use readyReplicas instead of conditions field.

**Comment by mikechengwei** [comment]
image

**Comment by mikechengwei** [comment]
> rest lgtm
> 
> could you show an example output?

image

File: pkg/monitor/monitor/monitor_manager.go [hunk 5]
```diff
@@ -170,9 +170,29 @@ func (m *MonitorManager) SyncMonitor(monitor *v1alpha1.TidbMonitor) error {
...
+
+	return nil
+}
+
+func (m *MonitorManager) syncTidbMonitorStatus(monitor *v1alpha1.TidbMonitor) error {
+	sts, err := m.deps.StatefulSetLister.StatefulSets(monitor.Namespace).Get(GetMonitorObjectName(monitor))
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			klog.V(4).Infof("tm[%s/%s]'s sts not found", monitor.Namespace, monitor.Name)
+		}
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:190]:
```suggestion
		if errors.IsNotFound(err) {
			klog.V(4).Infof("tm[%s/%s]'s sts not found", monitor.Namespace, monitor.Name)
			return nil
		}
```

File: pkg/util/crdutil.go [hunk 6]
```diff
@@ -296,6 +296,24 @@ var (
 		Type:     "date",
 		JSONPath: ".metadata.creationTimestamp",
 	}
+	tidbMonitorAdditionalPrinterColumns []extensionsobj.CustomResourceColumnDefinition
+
+	tidbMonitorDesireColumn = extensionsobj.CustomResourceColumnDefinition{
+		Name:        "Desire",
```
Comment (DanielZhangQD) [inline pkg/util/crdutil.go:302]:
```suggestion
		Name:        "DESIRED",
```

File: pkg/util/crdutil.go [hunk 7]
```diff
@@ -296,6 +296,24 @@ var (
 		Type:     "date",
 		JSONPath: ".metadata.creationTimestamp",
 	}
+	tidbMonitorAdditionalPrinterColumns []extensionsobj.CustomResourceColumnDefinition
+
+	tidbMonitorDesireColumn = extensionsobj.CustomResourceColumnDefinition{
```
Comment (DanielZhangQD) [inline pkg/util/crdutil.go:301]:
```suggestion
	tidbMonitorDesiredColumn = extensionsobj.CustomResourceColumnDefinition{
```

File: pkg/util/crdutil.go [hunk 8]
```diff
@@ -296,6 +296,24 @@ var (
...
+		Name:        "Desire",
+		Type:        "integer",
+		Description: "The desired replicas number of tidbMonitor cluster",
+		JSONPath:    ".spec.replicas",
+	}
+	readyTiDBMonitorAdditionalPrinterColumns = extensionsobj.CustomResourceColumnDefinition{
```
Comment (DanielZhangQD) [inline pkg/util/crdutil.go:307]:
```suggestion
	tidbMonitorReadyColumn = extensionsobj.CustomResourceColumnDefinition{
```

File: pkg/util/crdutil.go [hunk 9]
```diff
@@ -296,6 +296,24 @@ var (
...
+		Name:     "READY",
+		Type:     "string",
+		JSONPath: ".status.statefulSet.readyReplicas",
+	}
+	updateTiDBMonitorAdditionalPrinterColumns = extensionsobj.CustomResourceColumnDefinition{
```
Comment (DanielZhangQD) [inline pkg/util/crdutil.go:312]:
```suggestion
	tidbMonitorUpdatedColumn = extensionsobj.CustomResourceColumnDefinition{
```

File: pkg/monitor/monitor/monitor_manager.go [hunk 10]
```diff
@@ -170,9 +170,29 @@ func (m *MonitorManager) SyncMonitor(monitor *v1alpha1.TidbMonitor) error {
 		m.deps.Recorder.Event(monitor, corev1.EventTypeWarning, FailedSync, message)
 		return err
 	}
+
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:173]: No need to change this.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: pkg/util/crdutil.go (hunk 1 above)
Comment (mikechengwei) [reply pkg/util/crdutil.go]: I ignored this in the process of repeatedly optimizing the code, sorry.  I have changed it to the right way.

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**